src/types/ut_types/UTF8OutPipeReaderTests.cpp

```diff
@@ -0,0 +1,199 @@
+        {
+            ++successCounter;
+        }
+        ++subTestCounter;
+
+        // Test 4:
+        //                                                    |   |
+        utf8TestString.replace(bufferSize - 6, 12, "STUV\xF0\x90\x8D\x88WXYZ");
+        if (SUCCEEDED(RunTest(utf8TestString)))
+        {
+            ++successCounter;
+        }
+        ++subTestCounter;
+
+        // Test 5:
+        //                                                 |   |
+        utf8TestString.replace(bufferSize - 6, 12, "STUVW\xF0\x90\x8D\x88XYZ");
+        if (SUCCEEDED(RunTest(utf8TestString)))
+        {
+            ++successCounter;
+        }
+        ++subTestCounter;
+
+        // Test 6:
+        //                                               |  |
+        utf8TestString.replace(bufferSize - 6, 12, "STUVWX\xF0\x90\x8D\x88YZ");
+        if (SUCCEEDED(RunTest(utf8TestString)))
+        {
+            ++successCounter;
+        }
+        ++subTestCounter;
+
+        // Test 7:
+        //                                               ||
+        utf8TestString.replace(bufferSize - 6, 12, "STUVWXY\xF0\x90\x8D\x88Z");
+        if (SUCCEEDED(RunTest(utf8TestString)))
+        {
+            ++successCounter;
+        }
+        ++subTestCounter;
+
+        // Fail if not all of the sub-tests passed.
+        VERIFY_ARE_EQUAL(successCounter, subTestCounter);
+    }
+
+private:
+    struct ThreadData
+    {
+        HANDLE& writeTo;
+        std::string& utf8TestString;
+    };
+
+    // Thread function which writes the UTF-8 data to the pipe.
+    static DWORD WINAPI WritePipeThread(LPVOID threadArg)
+    {
+        ThreadData* pThreadData{ reinterpret_cast<ThreadData*>(threadArg) };
+        DWORD length{};
+
+        WriteFile(pThreadData->writeTo, pThreadData->utf8TestString.c_str(), static_cast<DWORD>(pThreadData->utf8TestString.size()), &length, nullptr);
+        CloseHandle(pThreadData->writeTo);
+
+        return 0;
+    }
+
+    // Performs the sub-tests.
+    HRESULT RunTest(std::string& utf8TestString)
+    {
+        HANDLE readFrom{ INVALID_HANDLE_VALUE }, writeTo{ INVALID_HANDLE_VALUE }; // pipe handles
+        SECURITY_ATTRIBUTES sa{ sizeof(SECURITY_ATTRIBUTES) };
+
+        std::string_view strView{}; // contains the chunk that we get from UTF8OutPipeReader::Read
+        const winrt::hstring utf16Expected{ winrt::to_hstring(utf8TestString) }; // contains the whole string converted to UTF-16
+        winrt::hstring utf16Actual{}; // will be concatenated from the converted chunks
+
+        CreatePipe(&readFrom, &writeTo, &sa, 0); // create the pipe handles
+
+        wil::unique_hfile outPipe{ readFrom };
+        static UTF8OutPipeReader reader{ outPipe }; // declare a static instance of UTF8OutPipeReader
+
+        ThreadData data{ writeTo, utf8TestString };
+
+        HANDLE threadHandle{ CreateThread(nullptr, 0, WritePipeThread, &data, 0, nullptr) }; // create a thread that writes to the pipe

```

Comment:
If you hold threadHandle in a wil::unique_handle, then it will automatically CloseHandle on it when the object goes out of scope and you can simplify some of your early-return-on-failures below to things like

RETURN_HR_IF_NULL(E_FAIL, threadHandle.get());